### PR TITLE
NFT Orderbook

### PR DESCRIPTION
**Pull request by bausano**

### CODE REVIEW

File: sources/trading/ob.move [hunk 1]
```diff
@@ -0,0 +1,909 @@
...
+    /// Offer given NFT to be traded for given (`requsted_tokens`) tokens. If
+    /// there exists a bid with higher offer than `requsted_tokens`, then trade
+    /// is immeidately executed. Otherwise the NFT is transferred to a newly
+    /// created ask object and the object is inserted to the orderbook.
+    public entry fun create_ask<C, FT>(
+        book: &mut Orderbook<C, FT>,
+        requsted_tokens: u64,
+        transfer_cap: TransferCap,
+        safe: &mut Safe,
+        whitelist: &Whitelist,
```
Comment (Suficio) [inline sources/trading/ob.move]: Are there any safety checks that `Whitelist` must pass, seems like anyone can pass their own whitelist that allows the transfer of NFTs by the OB?

**Review by Suficio — CHANGES_REQUESTED** [review]

File: sources/trading/ob.move (hunk 1 above)
Comment (bausano) [reply sources/trading/ob.move]: Yes, the collection creator must agree that their collection is added to the whitelist. If you create your own whitelist, OB won't be able to transfer the NFT unless SUIMARINES creators agreed to be part of your whitelist.

File: sources/trading/ob.move [hunk 2]
```diff
@@ -0,0 +1,912 @@
...
+        book: &mut Orderbook<C, FT>,
+        buyer_safe: &mut Safe,
+        price: u64,
+        wallet: &mut Coin<FT>,
+        ctx: &mut TxContext,
+    ) {
+        utils::assert_same_module_as_witness<C, W>();
+
+        create_bid_<C, FT>(book, buyer_safe, price, option::none(), wallet, ctx)
+    }
```
Comment (Suficio) [inline sources/trading/ob.move:215]: Spacing, in other places as well

File: sources/trading/ob.move [hunk 3]
```diff
@@ -0,0 +1,912 @@
...
+
+    public fun ask_nft(ask: &Ask): &TransferCap {
+        &ask.transfer_cap
+    }
+
+    public fun ask_owner(ask: &Ask): address {
+        ask.owner
+    }
+
+    fun create_<C, FT>(
```
Comment (Suficio) [inline sources/trading/ob.move:540]: Could this be `public fun new`?

File: sources/trading/ob.move [hunk 4]
```diff
@@ -0,0 +1,912 @@
...
+    ): Orderbook<C, FT> {
+        utils::assert_same_module_as_witness<C, W>();
+
+        create_<C, FT>(no_protection(), ctx)
+    }
+    public fun share<C, FT>(ob: Orderbook<C, FT>) {
+        share_object(ob);
+    }
+
+    public fun toggle_protection_on_buy_nft<W: drop, C, FT>(
```
Comment (Suficio) [inline sources/trading/ob.move:462]: Lets merge this as is, but would be good to make this explicit, `enable_protection_on_buy_nft`, `disable_protection_on_buy_nft`.

**Review by Suficio — APPROVED** [review]

File: sources/trading/ob.move [hunk 5]
```diff
@@ -0,0 +1,909 @@
...
+        // wallet
+        let bid_offer = balance::split(coin::balance_mut(wallet), price);
+
+        let asks = &mut book.asks;
+
+        let can_be_filled = !crit_bit::is_empty(asks) &&
+            crit_bit::min_key(asks) <= price;
+
+        if (can_be_filled) {
+            // OPTIMIZE: this is being recomputed
```
[the remark this answers is not in the document]
Comment (bausano) [reply sources/trading/ob.move]: done

**Pull request merged**